File: JS/src/next_permutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
function nextPermutation(nums) {
  let left = nums.length - 2;

  while (left >= 0 && nums[left] >= nums[left + 1]) {
    left--;
  }

  if (left < 0) {
    for (let i = 0; i < Math.trunc(nums.length / 2); i++) {
      let temp = nums[nums.length - 1 - i];
      nums[nums.length - 1 - i] = nums[i];
      nums[i] = temp;
    }
  } else {
    let right = nums.length - 1;
    while (nums[right] <= nums[left]) {
      right--;
    }

    let temp = nums[right];
    nums[right] = nums[left];
    nums[left] = temp;

    const rev = nums.splice(left + 1, nums.length - left - 1);
    nums.push(...rev.reverse());
  }
}

export { nextPermutation };

File: C++/src/next_permutation.cc
#include "next_permutation.hpp"
#include <algorithm>

void Solution::nextPermutation(vector<int> &nums) {
  size_t left = nums.size() - 2;

  while (int(left) >= 0 && nums[left] >= nums[left + 1]) {
    left--;
  }

  // If left is -1, it mean the current state is the last permutation
  if (int(left) == -1) {
    reverse(nums.begin(), nums.end());
  } else {
    size_t right = nums.size() - 1;
    while (nums[right] <= nums[left]) {
      right--;
    }

    swap(nums[left], nums[right]);

    reverse(nums.begin() + long(left) + 1, nums.end());
  }
}